### PR TITLE
Add support for direct play

### LINKED ISSUE #12: Direct Play
Support playing media files directly accessible to Kodi (e.g. on a local NAS) without going through the PMS.

This should be configurable i.e. the user should be able to choose on a media provider level whether Direct Play should be used or not.

### PR DESCRIPTION
**Pull request by JeordyR**
Update with direct play feature including settings:

* Enable/disable DirectPlayAllowed at the provider level
* Enable/disable path substitution at the import level
  * Source path to substitute
  * Substitution value

I threw a TODO [here](https://github.com/JeordyR/mediaimporter.plex/blob/efede4cc35e74584ea4e9eb19ce28e499d895001/plex/api.py#L391) to make sure we discussed this part.

I went through api, provider_observer, and observer and haven't had any luck figuring out how to get a mediaImport object to be able to pull in importSettings and providerSettings, or a way to get either of those without a mediaImport object at that point in the code.

I assume this will cause a gap where items added/updated via the websocket watcher will not end up getting the directPlay pathing that they should or having their directPlay pathing replaced with a stream URL if updating an existing item.

This setup is fully working with my server and test kodi setup with normal import. Using path substitution to replace `/data` with `smb://10.10.1.1/storage/media` to handle the pathing differences. Currently the DirectPlayAllowed provider settings is on by default, and fillVideoInfos will fall back to populating the streaming url if the direct_play pathing is not accessible. Wasn't sure if we would want the fall-back or throw an error if their direct_play config is wrong, so did the fall-back for now.

### CODE REVIEW

**Comment by Montellese** [comment]
Thanks for working on this. A few inputs:
* in `mediaimporter.emby` I only pass the relevant setting values (e.g. `allowDirectPlay`) into the helper methods and not the whole settings object.
* `mediaimporter.emby` has some special path handling code to map certain paths to what Kodi expects. See e.g. https://github.com/Montellese/mediaimporter.emby/blob/cb4b6d2246dcd2106e978298ae7dc320bc534d05/lib/kodi.py#L670.
* I'm not sure if your approach concerning path substitution is the right one. What if a Plex library pulls media items from multiple locations? With your approach only one path can be substituted. Kodi already has its own [path substitution](https://kodi.wiki/view/Path_substitution) implementation so I would not do this in the add-on but use what Kodi already provides.

**Comment by JeordyR** [comment]
> Thanks for working on this. A few inputs:
> 
> * in `mediaimporter.emby` I only pass the relevant setting values (e.g. `allowDirectPlay`) into the helper methods and not the whole settings object.
> * `mediaimporter.emby` has some special path handling code to map certain paths to what Kodi expects. See e.g. https://github.com/Montellese/mediaimporter.emby/blob/cb4b6d2246dcd2106e978298ae7dc320bc534d05/lib/kodi.py#L670.
> * I'm not sure if your approach concerning path substitution is the right one. What if a Plex library pulls media items from multiple locations? With your approach only one path can be substituted. Kodi already has its own [path substitution](https://kodi.wiki/view/Path_substitution) implementation so I would not do this in the add-on but use what Kodi already provides.

Gotchya, I didn't realize that the path substitution built-in worked that way. That will simplify things for this implementation.

I'm not sure how much of the path handling will be necessary with Plex since you are just deciding between the plex stream URL and Plex server's local path to the video file (plex doesn't support blu-ray folders or specifying non-local filesystem paths to worry about). Can basically remove all of the path substitution stuff I added, pass allowDirectPlay in, and then leave the existing stuff otherwise. If allowDirectPlay, check if Plex's file path is accessible (I assume the path substitution already defined in advancedsettings would be taken account of here). If accessible, store that as the path, and if not store the streaming url.

**Comment by JeordyR** [comment]
@Montellese Got it updated based on your comments. Tested and works well using Kodi's path substitution. Not needing importSettings anymore also found how to get the allowDirectPlay from provider_observer looking at the emby setup, so that gap should be filled as well.

**Comment by Montellese** [comment]
> I'm not sure how much of the path handling will be necessary with Plex since you are just deciding between the plex stream URL and Plex server's local path to the video file (plex doesn't support blu-ray folders or specifying non-local filesystem paths to worry about).

That's actually not true. Plex supports UNC paths on Windows but you can't automatically browse them. But if you manually copy the UNC path into the dialog it works perfectly fine. And Kodi can't handle those directly because it needs to be converted to `smb://`. So at least that part of the path mapping is also required here.

File: plex/api.py [hunk 1]
```diff
@@ -386,6 +391,8 @@ def getPlexItemAsListItem(
         if not plexItemId:
             raise ValueError('invalid plexItemId')
 
+        # TODO: How do we get the importSettings and providerSettings here?
```
Comment (Montellese) [inline plex/api.py]: Is this still valid? Maybe `allowDirectPlay` should be passed into `getPlexItemAsListItem()`?

File: plex/api.py [hunk 2]
```diff
@@ -663,13 +676,28 @@ def fillVideoInfos(
 
         if mediaPart:
             # extract the absolute / actual path and the stream URL from the selected MediaPart
-            info['path'] = mediaPart.file
-            item.setPath(plexServer.url(mediaPart.key, includeToken=True))
+            info['path'] = mediaPart.file  # path to file on disk
+
+            # determine if directPlay is enabled and possible
+            if allowDirectPlay:
+                log("Configuring item for direct play", xbmc.LOGDEBUG)
```
Comment (Montellese) [inline plex/api.py]: This log is not very helpful since it doesn't even mention the specific item. But I would drop it completely because it will generate a huge log spam when importing large libraries.

File: resources/providersettings.xml [hunk 3]
```diff
@@ -45,7 +45,7 @@
             <dependency type="enable">
               <or>
                 <condition on="setting" setting="plex.authentication" operator="is">0</condition>  <!-- SETTINGS_PROVIDER_AUTHENTICATION_OPTION_LOCAL -->
-                <condition on="setting" setting="plex.token" operator="!is"></condition> 
+                <condition on="setting" setting="plex.token" operator="!is"></condition>
```
Comment (Montellese) [inline resources/providersettings.xml]: This change looks unrelated. Can you move it into a separate "cosmetics" commit?

File: plex/api.py [hunk 4]
```diff
@@ -663,13 +676,28 @@ def fillVideoInfos(
...
+                    log(f"Direct play configured but path {filePath} is inaccessible.", xbmc.LOGWARNING)
+                    item.setPath(plexServer.url(mediaPart.key, includeToken=True))
+            else:
+                item.setPath(plexServer.url(mediaPart.key, includeToken=True))
         elif isFolder:
             # for folders use locations for the path
             if locations:
                 info['path'] = locations[0]
             item.setPath(plexServer.url(plexItem.key, includeToken=True))
+
```
Comment (Montellese) [inline plex/api.py]: This change looks unrelated. Can you move it into a separate "cosmetics" commit?

File: lib/settings.py [hunk 5]
```diff
@@ -110,7 +116,7 @@ def HaveChanged(importSettings: xbmcaddon.Settings, providerSettings: xbmcaddon.
             SynchronizationSettings.SaveHash(importSettings, newHash)
 
         return True
-    
+
```
Comment (Montellese) [inline lib/settings.py]: This change looks unrelated. Can you move it into a separate "cosmetics" commit?

**Review by Montellese — CHANGES_REQUESTED** [review]
Looks good apart from the TODO and the missing UNC path mapping.

File: plex/api.py (hunk 1 above)
Comment (JeordyR) [reply plex/api.py]: Yep, that isn't needed anymore. Got allowDirectPlay into that function and forgot to finish passing it through.

**Comment by JeordyR** [comment]
@Montellese Updates finished. Got the whitespaces that my IDE auto-removed added back in and brought _mapPath in to handle the UNC stuff. I had completely forgotten plex on Windows was a thing :P

File: plex/api.py [hunk 6]
```diff
@@ -665,8 +676,21 @@ def fillVideoInfos(
...
+            if allowDirectPlay:
+                # perform replacement in file path if configured
+                filePath = Api._mapPath(mediaPart.file)
+
+                # confirm file is accessible, fall back to streaming otherwise
+                if filePath and xbmcvfs.exists(filePath):
+                    item.setPath(filePath)
+                else:
+                    log(f"Direct play configured but path {filePath} is inaccessible.", xbmc.LOGWARNING)
+                    item.setPath(plexServer.url(mediaPart.key, includeToken=True))
```
Comment (Montellese) [inline plex/api.py]: This line appears twice now. Maybe don't do anything here and in the `else` below and then check if `item.getPath()` is empty and fall back to using `item.setPath(plexServer.url(mediaPart.key, includeToken=True))` then to avoid code duplication.

**Review by Montellese — COMMENTED** [review]
One more minor and then this is good to go. Thanks!

**Comment by JeordyR** [comment]
> One more minor and then this is good to go. Thanks!

Updated, should be good to.

**Review by Montellese — APPROVED** [review]

**Comment by Montellese** [comment]
Argh almost forgot :-) Can you add a line to `changelog.txt` mentioning the added feature. I did this for your recent pull requests because I also forgot it there.

**Comment by JeordyR** [comment]
> Argh almost forgot :-) Can you add a line to `changelog.txt` mentioning the added feature. I did this for your recent pull requests because I also forgot it there.

Changelog updated.

**Review by Montellese — APPROVED** [review]

**Comment by Montellese** [comment]
Thanks for your patience.
